Automattic/vip-cli
# Add import files validation

##  Description

Basic validator to validate media files for imports.

Checking for:

1. Folder structure/path validation
    - Single sites:
      - There’s an `uploads` directory
      - Recommended: There’s a year directory within the `uploads` directory
      - Recommended: There’s a month directory within the year directory
      - Recommended path: `uploads/2020/6/picname.jpg`
    - Not checking for multisites at the moment

2. File media types (e.g. – png, jpg, etc.)
    - Check for missing or invalid file types

3. Filename character validation
    - What's allowed:
      - Non-English characters
      - Special characters: `spaces ( ) [ ] ~`
    - What's not allowed:
      - Special characters: `+ & # % = \' \" \ < > : ; , / ? $ * | ` ! { }`

4. Intermediate images
Detect intermediate images, such as:

`panda4000x6000.jpg` (sizing w/ no space/dash/underscore)
`panda-4000x6000.jpg` (dash)
`panda_4000x6000.jpg` (underscore)
`panda test 4000x6000.jpg` (spaces)
`panda_test-4000x6000@2x.jpg` (retina display)

## Steps to Test

1. Check out PR.
1. Run `npm run jest lib/vip-import-validate-files.js`
1. All tests should pass
1. Download this sample `uploads` directory and unzip: 
[uploads.zip](https://github.com/Automattic/vip/files/5014552/uploads.zip)

1. Run `npm install`
1. Run `npm run build`
1. Run `node ./dist/bin/vip-import-validate-files.js uploads`
1. Verify messages:

<img width="1788" alt="validation" src="https://user-images.githubusercontent.com/29221840/88997613-eebc2a80-d2a4-11ea-971f-3170962e3259.png">

Other test folders:
[folder1.zip](https://github.com/Automattic/vip/files/5004948/folder1.zip)

<img width="1182" alt="validation2" src="https://user-images.githubusercontent.com/29221840/88997745-3d69c480-d2a5-11ea-80b7-5c8ebd3fc3a4.png">
